Refactors automotive_demo to increase road network selection scalability.

liangfok opened the pull request:
This PR contains no functional changes. It simply refactors `automotive_demo` to allow more types of road networks to be supported in the future. Specifically, a demo is initialized by first loading the road network and then loading the vehicles. In addition, a new command line argument called `--road_network` is added to allow the user to select which type of road network to load.

The recommended review order is to start with the `main()` method and then follow the various methods that it calls.

It was extracted from #5185.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5233)
<!-- Reviewable:end -->

liangfok commented:
+@maddog-tri for feature review please.
+@jwnimmer-tri for platform review please.

Thanks.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-:-KdShRnPlQZQD08GDIhm:b-kigsme)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSiH4yrxgKrShgLVZR:-KdSiH4yrxgKrShgLVZS:bazmd1u) ([raw file](https://github.com/robotlocomotion/drake/blob/1d159e3e9016aa5e85f948c5f6678aaf99e64458/drake/automotive/automotive_demo.cc#L56)):*
> ```cpp
>   kMonolane = 1
> };
> ```

_Meta_ integers as a UX is obscure and brittle.  Let's use `string`.

---

*[drake/automotive/BUILD, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSipNIb3IoLL347Nuh:-KdSipNIb3IoLL347Nui:bpyyae8) ([raw file](https://github.com/robotlocomotion/drake/blob/1d159e3e9016aa5e85f948c5f6678aaf99e64458/drake/automotive/BUILD#L300)):*
> ```Python
>     ],
>     args = [
>         "--road_network=1",
> ```

_Meta_ Why doesn't the demo program know that this should be the default due to either `-road_file` or `-road_path`?  Why brittlely have to put a `1` in the BUILD files?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/automotive/automotive_demo.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSiH4yrxgKrShgLVZR:-KdStIyWcUIWWIfdSAfO:b-goni66) ([raw file](https://github.com/robotlocomotion/drake/blob/1d159e3e9016aa5e85f948c5f6678aaf99e64458/drake/automotive/automotive_demo.cc#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ integers as a UX is obscure and brittle.  Let's use `string`.
</blockquote></details>

Done (removed) from public UX. It's now automatically determined. See `DetermineRoadNetworkType()`.

---

*[drake/automotive/BUILD, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSipNIb3IoLL347Nuh:-KdSt3l5YjnOcClEY6dP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1d159e3e9016aa5e85f948c5f6678aaf99e64458/drake/automotive/BUILD#L300)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ Why doesn't the demo program know that this should be the default due to either `-road_file` or `-road_path`?  Why brittlely have to put a `1` in the BUILD files?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r2.
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 31 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxLaauMTpqGYW9Fwp:-KdSxLabFPK9kGXM13kK:b-eyg7uw) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L31)):*
> ```cpp
> DEFINE_int32(num_ado_car, 1,
>     "Number of vehicles controlled by a (possibly trivial) traffic model");
> DEFINE_string(simple_car_names, "",
> ```

Dead param.

---

*[drake/automotive/automotive_demo.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxNTqsdbUAYi4ayPx:-KdSxNTqsdbUAYi4ayPy:b-eyg7uw) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L35)):*
> ```cpp
>     "cars subscribed to DRIVING_COMMAND_Russ, "
>     "DRIVING_COMMAND_Jeremy, and DRIVING_COMMAND_Liang)");
> DEFINE_int32(num_trajectory_car, 1,
> ```

Dead param.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 31 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxLaauMTpqGYW9Fwp:-KdT-It3AyA7KuIn6EsW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Dead param.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxNTqsdbUAYi4ayPx:-KdT-M9WQek_EEzsqbsC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Dead param.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxcRNVQuI_2YQ4RZw:-KdSxcRNVQuI_2YQ4RZx:b-6dz6my) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L38)):*
> ```cpp
>     "Number of TrajectoryCar vehicles");
> 
> // Simulation parameters.
> ```

I don't think this comment (and the "// Vehicle parameters." help much, and they will bitrot.  For example `use_idm` below is a vehicle param, not a simulation / world parameter.  If a developer can't instantaneously know how the parameter will be used with at most simple Ctrl-F then the code is already too obscure and comments like this will not help.

---

*[drake/automotive/automotive_demo.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSy2TuXGJXBgR8yJVD:-KdSy2TuXGJXBgR8yJVE:by02cr8) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L52)):*
> ```cpp
>               "lane id must be prefixed by either 'start:' or 'end:' "
>               "indicating at which end of the first lane to begin the "
>               "circuit.  If the string is empty, a default path will "
> ```

_Meta_ If we are going to rewrap the param strings in this PR, we might as well make them all consistent so that future authors have some hope of keeping the indentation choice intact.

---

*[drake/automotive/automotive_demo.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSydCammeI1nG2btYT:-KdSydCammeI1nG2btYU:b3j86t6) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L66)):*
> ```cpp
>   kFlatTerrain = 0,
>   kMonolane = 1
> };
> ```

Brittle.  Use `enum class` and don't use `int` in the below code.  The integer values should never be meaningful, and we should embrace the type checking that `enum class` offers.

---

*[drake/automotive/automotive_demo.cc, line 73 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSymgAaFlBjEekmNRY:-KdSymgAaFlBjEekmNRZ:b-ha8a7z) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L73)):*
> ```cpp
> // geometry. For exeample, the call to CreateTrajectoryParams() below
> // expects a "car" to have a particular length and width.
> const char* kSdfFile =
> ```

`const char* const`

---

*[drake/automotive/automotive_demo.cc, line 255 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzvHMnSJS9k8dH7QB:-KdSzvHNeLzql1Thm0op:b9zv7ud) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L255)):*
> ```cpp
> // or 'end:'.
> const maliput::api::RoadGeometry*
>     AddMonolane(AutomotiveSimulator<double>* simulator) {
> ```

_Meta_ Bad indentation.  Function name either left-aligns, or appears after the return type.  Only parameters are +4 indent.

---

*[drake/automotive/automotive_demo.cc, line 258 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdT-ICCWpc6aRHF1Tm1:-KdT-ICCWpc6aRHF1Tm2:bkp7mmo) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L258)):*
> ```cpp
>   if (FLAGS_road_file.empty()) {
>     drake::log()->error("AddMonolane(): No monolane road file specified.");
>     DRAKE_ABORT();
> ```

BTW _Meta_ Consider `DRAKE_ABORT_MSG`.

---

*[drake/automotive/automotive_demo.cc, line 309 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzYDIBDBSdsDY2bFy:-KdSzYDIBDBSdsDY2bFz:byqqvfc) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L309)):*
> ```cpp
>     case kFlatTerrain:
>       AddFlatTerrain(simulator);
>     break;
> ```

_Meta_ wrong indent.  _Meta_ `switch-case` bodies require mandatory curly braces.

---

*[drake/automotive/automotive_demo.cc, line 316 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzhZCfkX1p98cZK7M:-KdSzhZCfkX1p98cZK7N:bwwyb1m) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L316)):*
> ```cpp
>       drake::log()->error("AddTerrain(): Unsupported road network type: {}.",
>           std::to_string(road_network_type));
>       DRAKE_ABORT();
> ```

This won't be needed once we have `enum class`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/automotive_demo.cc, line 97 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdT03I5Ro-8dU-j3m6L:-KdT03I5Ro-8dU-j3m6M:b-28blz) ([raw file](https://github.com/robotlocomotion/drake/blob/67c533ab8362a5dde59068b0d177f2ffbef53ada/drake/automotive/automotive_demo.cc#L97)):*
> ```cpp
> }
> 
> void AddMonolaneVehicles(
> ```

It is possible to relocate this code to more like where it used to be in master, so that the diff shows up better?  We can always shuffle functions to be more logically ordered later.  If not just let me know, and I will fire up some offline diff tools instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233)*
<!-- Sent from Reviewable.io -->

liangfok commented:
While I realize this PR will likely need to be put on pause while #5167 is being reprocessed.

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSxcRNVQuI_2YQ4RZw:-KdT5fyDRerivVQnz2Cd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think this comment (and the "// Vehicle parameters." help much, and they will bitrot.  For example `use_idm` below is a vehicle param, not a simulation / world parameter.  If a developer can't instantaneously know how the parameter will be used with at most simple Ctrl-F then the code is already too obscure and comments like this will not help.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSy2TuXGJXBgR8yJVD:-KdT61nb6q70Zyvipkji:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ If we are going to rewrap the param strings in this PR, we might as well make them all consistent so that future authors have some hope of keeping the indentation choice intact.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSydCammeI1nG2btYT:-KdT6TdvdVv7xHLoU0id:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Brittle.  Use `enum class` and don't use `int` in the below code.  The integer values should never be meaningful, and we should embrace the type checking that `enum class` offers.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 73 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSymgAaFlBjEekmNRY:-KdT6YZyhHdJxYc3qG2M:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L73)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`const char* const`
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 255 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzvHMnSJS9k8dH7QB:-KdT6uGsdd085noJ7zTS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L255)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ Bad indentation.  Function name either left-aligns, or appears after the return type.  Only parameters are +4 indent.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 258 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdT-ICCWpc6aRHF1Tm1:-KdT7-TANgJ9Zq3RbF7B:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L258)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW _Meta_ Consider `DRAKE_ABORT_MSG`.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 309 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzYDIBDBSdsDY2bFy:-KdT7iH-XhkW4F4f6P1x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L309)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ wrong indent.  _Meta_ `switch-case` bodies require mandatory curly braces.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 316 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdSzhZCfkX1p98cZK7M:-KdT7mJ7hYegpr6bxPw4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/39c49ecab3572fe2c78b38948173a8b7107af3de/drake/automotive/automotive_demo.cc#L316)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This won't be needed once we have `enum class`.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 97 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-KdT03I5Ro-8dU-j3m6L:-KdT8xsmwqvVPYm7Jqcf:b-p16h53) ([raw file](https://github.com/robotlocomotion/drake/blob/67c533ab8362a5dde59068b0d177f2ffbef53ada/drake/automotive/automotive_demo.cc#L97)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It is possible to relocate this code to more like where it used to be in master, so that the diff shows up better?  We can always shuffle functions to be more logically ordered later.  If not just let me know, and I will fire up some offline diff tools instead.
</blockquote></details>

I moved it closer to where it was before. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-:-KdT9CjSQGfDkmUhf-tS:b-hbkkgs)*
<!-- Sent from Reviewable.io -->

liangfok commented:
...I addressed all reviewer comments.

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5233#-:-KdT9Xwrua8su8pgwIDQ:b2hguaz)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm applying the "do not review" label until after #5235 is merged.

liangfok commented:
This PR is no longer necessary due to #5235.
